src/inmanta/env.py

```diff
@@ -560,39 +564,87 @@ def init_env(self) -> bool:
             except Exception:
-                LOGGER.exception("Unable to create new virtualenv at %s", self.env_path)
-                return False
+                raise VenvCreationFailedError(msg=f"Unable to create new virtualenv at {self.env_path}")
             LOGGER.debug("Created a new virtualenv at %s", self.env_path)
+        elif not os.path.exists(self._path_sitecustomize_py):
+            # Venv was created using an older version of Inmanta -> Update pip binary and set sitecustomize.py file
+            self._write_pip_binary()
+            self._write_sitecustomize_py_file()
 
         # set the path to the python and the pip executables
         self.virtual_python = self.python_path
 
-        return True
-
     def is_using_virtual_env(self) -> bool:
         return self.__using_venv
 
     def use_virtual_env(self) -> None:
         """
-        Use the virtual environment
+        Activate the virtual environment.
         """
         if self.__using_venv:
             raise Exception(f"Already using venv {self.env_path}.")
 
-        if not self.init_env():
-            raise Exception("Unable to init virtual environment")
-
+        self.init_env()
         self._activate_that()
 
         # patch up pkg
         self.notify_change()
 
         self.__using_venv = True
 
+    def _write_pip_binary(self) -> None:
+        """
+        write out a "stub" pip binary so that pip list works in the virtual env.
+        """
+        pip_path = os.path.join(self.env_path, "bin", "pip")
+
+        with open(pip_path, "w", encoding="utf-8") as fd:
+            fd.write(
+                """#!/bin/bash
+cd $(dirname "${{BASH_SOURCE[0]}}")
+source activate
+python -m pip $@
+                """.strip()
+            )
+        os.chmod(pip_path, 0o755)
+
+    def _write_sitecustomize_py_file(self) -> None:
+        """
+        Write a sitecustomize.py file to the venv to ensure that an activation of this venv will also activate
+        the parent venv. The site directories of the parent venv should appear later in sys.path than the ones of this venv.
+        """
+        sys_path_as_python_strings = ['"' + p.replace('"', '\\"') + '"' for p in self._get_sys_path_after_activation()]
+        script = f"""import sys
+import site
+previous_sys_path = list(sys.path)
+site_dirs = [{', '.join(sys_path_as_python_strings)}]
+for s in site_dirs:
+    site.addsitedir(s)
+# Move the added items to the front of the path
+new_entries_sys_path = [e for e in sys.path if e not in previous_sys_path]
+sys.path = [*new_entries_sys_path, *previous_sys_path]
+"""
+        with open(self._path_sitecustomize_py, "w", encoding="utf-8") as fd:
+            fd.write(script)
+
+    def _get_sys_path_after_activation(self) -> List[str]:
+        """
+        Returns the content that sys.path should have after activating this venv.
+        """
+        prev_sys_path = list(sys.path)
```

Comment:
this bit I don't understand for `_write_sitecustomize_py_file` 

1. This looks like a huge side effect?, addsitedir executes pth files iirc, so this could do anything to this python interpreter?
2. why not just add the current venv's site_packages folder to the site_dir?

Reply:
I did this to make sure that both the parent and the child venv remain in sync all the time. With this implementation, an editable install in the parent venv will be seen by the child venv. If we only add current venv's site dir, this is not the case. But this functionality is not strictly required for the use case we consider here. I will make sure that we only add the current venv's site dir.

Reply:
if we use `addsitedir` then all pth files are executed I think, so editable installs should even work?